inference-engine/src/inference_engine/src/threading/ie_parallel_custom_arena.cpp

```diff
@@ -73,37 +74,87 @@ static bool is_binding_environment_valid() {
-
-static int core_types_count = 0;
-static int* core_types_indexes = nullptr;
-
-static void initialize_system_topology() {
-    static std::once_flag is_topology_initialized;
-
-    std::call_once(is_topology_initialized, [&] {
-        if (is_binding_environment_valid()) {
+class TBBSystemTopology {
+    TBBSystemTopology() {
+        if (binding_environment_valid) {
             __TBB_internal_initialize_system_topology(get_processors_group_num(),
                                                       numa_nodes_count,
                                                       numa_nodes_indexes,
                                                       core_types_count,
                                                       core_types_indexes);
-        } else {
-            static int dummy_index = task_arena::automatic;
+        }
+    }
 
-            numa_nodes_count = 1;
-            numa_nodes_indexes = &dummy_index;
+public:
+    ~TBBSystemTopology() {
+        if (binding_environment_valid) {
+            __TBB_internal_destroy_system_topology();
+        }
+    }
 
-            core_types_count = 1;
-            core_types_indexes = &dummy_index;
+    std::vector<numa_node_id> numa_nodes() const {
+#        if USE_TBBBIND_2_5
+        std::vector<numa_node_id> node_indexes(numa_nodes_count);
+        std::memcpy(node_indexes.data(), numa_nodes_indexes, numa_nodes_count * sizeof(int));
+        return node_indexes;
+#        elif TBB_NUMA_SUPPORT_PRESENT
+        return tbb::info::numa_nodes();
+#        else
+        return {tbb::task_arena::automatic};
+#        endif
+    }
+
+    std::vector<core_type_id> core_types() const {
+#        if USE_TBBBIND_2_5
+        std::vector<numa_node_id> core_type_indexes(core_types_count);
+        std::memcpy(core_type_indexes.data(), core_types_indexes, core_types_count * sizeof(int));
+        return core_type_indexes;
+#        elif TBB_HYBRID_CPUS_SUPPORT_PRESENT
+        return tbb::info::core_types();
+#        else
+        return {tbb::task_arena::automatic};
+#        endif
+    }
+
+    int default_concurrency(task_arena::constraints c) const {
+        if (c.max_concurrency > 0) {
+            return c.max_concurrency;
         }
-    });
+#        if USE_TBBBIND_2_5
+        if (binding_environment_valid) {
+            return __TBB_internal_get_default_concurrency(c.numa_id, c.core_type, c.max_threads_per_core);
+        }
+        return tbb::this_task_arena::max_concurrency();
+#        elif TBB_HYBRID_CPUS_SUPPORT_PRESENT
+        return tbb::info::default_concurrency(convert_constraints(c));
+#        elif TBB_NUMA_SUPPORT_PRESENT
+        return tbb::info::default_concurrency(c.numa_id);
+#        else
+        return tbb::this_task_arena::max_concurrency();
+#        endif
+    }
+
+    friend const TBBSystemTopology& system_topology();
+
+private:
+    int dummy_index = task_arena::automatic;
+    bool binding_environment_valid = is_binding_environment_valid();
```

Comment:
Why do you prefer to use bool variable instead of the function?

Reply:
I have cached the function result and avoided the additional system calls.

Reply:
I don't think so, the result is already cached inside the function by the static variable. It would be thread-safely initialized during the first call of the function and then just used as-is. So I think the separate variable is redundant.

Reply:
oh, I didn't notice static variable inside is_binding_environment_valid.  Ok, I'll temporary boolean variable.